projects/driver_controls/inc/ads1015_def.h

```diff
@@ -0,0 +1,108 @@
+#define ADS1015_PGA_FSR_2048 ((uint8_t)0x2 << 1)  // ±2.048 V
+#define ADS1015_PGA_FSR_1024 ((uint8_t)0x3 << 1)  // ±1.024 V
+#define ADS1015_PGA_FSR_512 ((uint8_t)0x4 << 1)   // ±0.512 V
+#define ADS1015_PGA_FSR_256 ((uint8_t)0x5 << 1)   // ±0.256 V
+
+// Conversion mode (single or continuous)
+#define ADS1015_CONVERSION_MODE_CONT ((uint8_t)0x0)
+#define ADS1015_CONVERSION_MODE_SINGLE ((uint8_t)0x1)  // default
+
+// Data sampling rate of the adc
+#define ADS1015_DATA_RATE_128 ((uint8_t)0x0 << 5)   //  128 SPS
+#define ADS1015_DATA_RATE_250 ((uint8_t)0x1 << 5)   //  250 SPS
+#define ADS1015_DATA_RATE_490 ((uint8_t)0x2 << 5)   //  490 SPS
+#define ADS1015_DATA_RATE_920 ((uint8_t)0x3 << 5)   //  920 SPS
+#define ADS1015_DATA_RATE_1600 ((uint8_t)0x4 << 5)  // 1600 SPS
+#define ADS1015_DATA_RATE_2400 ((uint8_t)0x5 << 5)  // 2400 SPS
+#define ADS1015_DATA_RATE_3300 ((uint8_t)0x6 << 5)  // 3300 SPS
+
+// Comparator operating mode
+#define ADS1015_COMP_MODE_TRAD ((uint8_t)0x0 << 4)  // default
+#define ADS1015_COMP_MODE_WINDOW ((uint8_t)0x1 << 4)
+
+// Polarity of the ALERT/RDY pin (active low or active high)
+#define ADS1015_COMP_POL_LOW ((uint8_t)0x0 << 3)
+#define ADS1015_COMP_POL_HIGH ((uint8_t)0x1 << 3)
+
+// Latching comparator
+#define ADS1015_COMP_LAT_NON_LATCHING ((uint8_t)0x0 << 2)
+#define ADS1015_COMP_LAT_LATCHING ((uint8_t)0x1 << 2)
+
+// Comparator queue and disable
+#define ADS1015_COMP_QUE_1_CONV ((uint8_t)0x0)  // Assert after one conversion
+#define ADS1015_COMP_QUE_2_CONV ((uint8_t)0x1)  // Assert after two conversion
+#define ADS1015_COMP_QUE_4_CONV ((uint8_t)0x2)  // Assert after four conversion
+#define ADS1015_COMP_QUE_DISABLE_COMP \
+  ((uint8_t)0x3)  // Disable comparator and set ALERT/RDY pin to high-impedance
+
+// **************************************************************************************
+
+// Setup for the config register's upper byte
+#define ADS1015_CONFIG_REGISTER_MSB(channel)                                 \
+  (ADS1015_START_SINGLE_CONV | ADS1015_AIN(channel) | ADS1015_PGA_FSR_4096 | \
+   ADS1015_CONVERSION_MODE_SINGLE)
+
+// Setup for the config register's lower byte
+#define ADS1015_CONFIG_REGISTER_LSB                                          \
+  (ADS1015_DATA_RATE_1600 | ADS1015_COMP_MODE_TRAD | ADS1015_COMP_POL_HIGH | \
+   ADS1015_COMP_LAT_NON_LATCHING | ADS1015_COMP_QUE_1_CONV)
+
+// These represent the full-scale range of ADS1015 scaling in mVolts.
+// They are used for calculating the LSB size, corresponding to PGA settings.
+// From section 8.3.3 of the datasheet.
+// Mult. by 2 corresponds to the range from -FS to +FS.
+#define ADS1015_FSR_6144 6144 * 2
+#define ADS1015_FSR_4096 4096 * 2
+#define ADS1015_FSR_2048 2048 * 2
+#define ADS1015_FSR_1024 1024 * 2
+#define ADS1015_FSR_512 512 * 2
+#define ADS1015_FSR_256 256 * 2
+
+// The LSB size in mVolt from section 8.3.3 of the datasheet.
+#define ADS1015_LSB_SIZE(fsr) fsr / (1 << 12)
```

Comment:
When using macros, you should always wrap arguments in parentheses. See [link](https://docs.freebsd.org/info/cpp/cpp.info.Macro_Parentheses.html) for information.

Reply:
Won't this result in non-integer LSBs?

Reply:
Ah that was a very nice link! 

I will cast it into int.

Reply:
It isn't that you should be casting it - it will be an int by default. It's that because it's an integer, trying to divide by a larger number will just result in 0.

i.e. `256 * 2 / (1 << 12) = 0`

The idea behind adding parentheses is this: Imagine if your argument was `6144 + 5`. Then, you'd get `6144 + 5 / (1 << 12) = 6144 + 5/4096 = 6144` which isn't what you intended.

Reply:
Yah I understood that part about the brackets. 
The problem with the division is a different problem though, I'm guessing. 
I have to change the order of the math.

Reply:
Yeah, exactly. It's because integer division truncates towards 0. The idea is to order your operations so that you're dividing a numerator larger than the denominator.